### PR TITLE
KAFKA-12508: Emit records with same value and same timestamp

### PR DESCRIPTION
**Pull request by cadonna**
Emit on change introduced in Streams with KIP-557 might lead to
data loss if a record is put into a source KTable and emitted
downstream and then a failure happens before the offset could be
committed. After Streams rereads the record, it would find a record
with the same key, value and timestamp in the KTable (i.e. the same
record that was put into the KTable before the failure) and not
forward it downstreams. Hence, the record would never be processed
downstream of the KTable which breaks the at-least-once
processing guarantee.

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Comment by cadonna** [comment]
Call for review: @vvcephei, @ableegoldman, @mimaison

File: streams/src/main/java/org/apache/kafka/streams/state/internals/ValueAndTimestampSerializer.java
Comment (ableegoldman) [inline streams/src/main/java/org/apache/kafka/streams/state/internals/ValueAndTimestampSerializer.java:108]: This is a bit unfortunate since it essentially breaks the emit-on-change semantics 😕  I guess it should be relatively rare for no-op updates to come in with the same timestamp, but this still seems like kind of a structural failure of Kafka Streams. We shouldn't need to assume that if we find an identical record in the state store then we have to forward it on the off-chance it was (1) actually the same record, and (2) was only partially processed when we happened across an unexpected exception.

I'm not saying I have a better idea that could be implemented quickly & safely given the releases we're blocking, but we should at least file a ticket so users are aware of this flaw. Otherwise I imagine we might be getting bug reports that emit-on-change doesn't work. It's possible some of the orthogonal work that's been discussed in the past will end up fixing this on the side (eg buffering updates before commit, versioned tables, etc), whenever we finally get around to any of that

**Review by ableegoldman — APPROVED** [review]
Thanks for the PR! It's kind of a bummer that we have to do this, but getting a few extra updates is better than dropping records on the floor.

**Comment by ableegoldman** [comment]
One unrelated test failure: `kafka.server.ListOffsetsRequestTest.testResponseIncludesLeaderEpoch()`

Merging now

**Comment by ableegoldman** [comment]
Merged to trunk and cherrypicked back to 2.6 & 2.7. Waiting for @vvcephei to approve this blocker before merging to 2.8

The integration test was relying on the new uncaught exception handler which does not exist prior to 2.8, so I replaced that with just using two StreamThreads. This should have the exact same effect

**Comment by ableegoldman** [comment]
Cherrypicked to 2.8 cc @vvcephei

**Pull request merged**
